homeassistant/components/rympro/coordinator.py

```diff
@@ -29,11 +32,50 @@ def __init__(self, hass: HomeAssistant, rympro: RymPro) -> None:
             name=DOMAIN,
             update_interval=interval,
         )
+        self._meter_sensors: dict[int, set[MeterSensor]] = {}
+        self._first_run = True
 
-    async def _async_update_data(self) -> dict[int, dict]:
+    def add_meter_sensor(
+        self, meter_id: int, sensor: MeterSensor
+    ) -> Callable[[], None]:
+        """Add a meter sensor to be fetched."""
+        self._meter_sensors.setdefault(meter_id, set())
+        self._meter_sensors[meter_id].add(sensor)
+
+        def _remove():
+            self._meter_sensors[meter_id].remove(sensor)
+
+        return _remove
+
+    async def _sensor_call(self, meter_id: int, sensor: MeterSensor) -> Any:
+        return await getattr(self.rympro, sensor.value)(meter_id)
+
+    async def _results_for_meter(self, meter_id: int) -> dict[MeterSensor, Any]:
+        if self._first_run:
+            sensors = set(MeterSensor)
+        else:
+            sensors = self._meter_sensors.get(meter_id, set())
+        sensors.discard(MeterSensor.TOTAL_CONSUMPTION)
+        values = await asyncio.gather(
+            *(self._sensor_call(meter_id, sensor) for sensor in sensors)
+        )
+        return dict(zip(sensors, values))
+
+    async def _async_update_data(self) -> dict[int, dict[MeterSensor, Any]]:
```

Comment:
The changes to `_async_update_data` do not exactly make the code easier to read!
I keep getting lost as I try to follow it...

Reply:
I tried to break down into smaller functions. Let me know if you think that helps

Reply:
I think you should create a dedicate model to store the `MeterData`, with four properties.

```suggestion
    async def _async_update_data(self) -> dict[int, MeterData]:
```